File: 백준/Gold/2636. 치즈/치즈.py
# 백준 2636
# 치즈가 모두 녹아 없어지는 데 걸리는 시간
# 모두 녹기 한 시간 전에 남아있는 치즈조각이 놓여 있는 칸의 개수
from collections import deque

n, m = map(int, input().split())

graph = []
for i in range(n):
    graph.append(list(map(int, input().split())))

dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]
result = []


def bfs():
    q = deque()
    q.append([0, 0])
    visited[0][0] = 1
    cnt = 0

    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < m and visited[nx][ny] == 0:
                if graph[nx][ny] == 0:
                    visited[nx][ny] = 1
                    q.append([nx, ny])
                elif graph[nx][ny] == 1:
                    graph[nx][ny] = 0
                    visited[nx][ny] = 1
                    cnt += 1
    result.append(cnt)
    return cnt


time = 0
while True:
    time += 1
    visited = [[0] * m for _ in range(n)]
    count = bfs()
    if count == 0:
        break

print(time - 1)
print(result[-2])

File: 백준/Gold/2638. 치즈/치즈.js
const path = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(path).toString().trim().split("\n");

class Queue {
  constructor() {
    this.items = {};
    this.headIndex = 0;
    this.tailIndex = 0;
  }
  enqueue(item) {
    this.items[this.tailIndex] = item;
    this.tailIndex++;
  }
  dequeue() {
    const item = this.items[this.headIndex];
    delete this.items[this.headIndex];
    this.headIndex++;
    return item;
  }
  peek() {
    return this.items[this.headIndex];
  }
  getLength() {
    return this.tailIndex - this.headIndex;
  }
}

const solution = (input) => {
  const [n, m] = input[0].split(' ').map(Number);
  const graph = [];
  for (let i = 1; i <= n; i += 1) {
    graph.push(input[i].split(' ').map(Number));
  }
  const dx = [0, 0, 1, -1];
  const dy = [1, -1, 0, 0];

  const bfs = () => {
    let visited = [];
    for (let i = 0; i < n; i += 1) {
      visited.push(new Array(m).fill(false));
    }
    visited[0][0] = true;
    const queue = new Queue();
    queue.enqueue([0, 0]);
    while (queue.getLength() !== 0) {
      const [x, y] = queue.dequeue();
      for (let i = 0; i < 4; i += 1) {
        const nx = x + dx[i];
        const ny = y + dy[i];
        if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
        if (!visited[nx][ny]) {
          if (graph[nx][ny] >= 1) graph[nx][ny] += 1;
          else {
            visited[nx][ny] = true;
            queue.enqueue([nx, ny]);
          }
        }
      }
    }
  }

  const melt = () => {
    let finish = true;
    for (let i = 0; i < n; i += 1) {
      for (let j = 0 ; j < m; j += 1) {
        if (graph[i][j] >= 3) {
          graph[i][j] = 0;
          finish = false;
        }
        else if (graph[i][j] === 2) graph[i][j] = 1;
      }
    }
    return finish;
  }

  let result = 0;
  while (true) {
    bfs();
    if (melt()) break;
    else result += 1;
  }
  return result;
}

const answer = solution(input);
console.log(answer);
